File: 数据结构/Python/第3章 链表/3.5 逆序输出链表/inverse.py
import node
import linked_list
import stack

def inverse(node):
    if node:
        inverse(node.next)
        print(node.data, end=' ')

def inverse_non_recursive(lst):
    s = stack.Stack()

    node = lst.get_head()
    while node:
        s.push(node)
        node = node.next
    while not s.is_empty():
        print(s.pop().data, end=' ')
    print()

def main():
    lst = linked_list.LinkedList()

    n = 6
    for i in range(n):
        lst.insert(i, node.Node(i))
    
    inverse(lst.get_head())
    print()

    inverse_non_recursive(lst)

if __name__ == "__main__":
    main()

File: 数据结构/Java/第3章 链表/3.5 逆序输出链表/Inverse.java
import java.util.ArrayList;

public class Inverse {
    public static void main(String[] args) {
        LinkedList list = new LinkedList();

        int n = 6;
        for(int i = 0; i < n; i++) {
            list.insert(i, new Node(i));
        }

        inverse(list.getHead());
        System.out.println();

        inverseNonRecursive(list);
    }

    public static void inverse(Node head) {
        if(head != null) {
            inverse(head.next);
            System.out.print(head.data + " ");
        }
    }

    public static void inverseNonRecursive(LinkedList list) {
        Stack s = new Stack();

        Node node = list.getHead();
        while(node != null) {
            s.push(node);
            node = node.next;
        }
        while(!s.isEmpty()) {
            System.out.print(s.pop().data + " ");
        }
        System.out.println();
    }
}
